### PR TITLE
[RFC] Initial support for building druntime with musl libc

### PR DESCRIPTION
**Pull request by yshui**
Make druntime buildable with CRuntime_Musl defined instead of CRuntime_Glibc.

### CODE REVIEW

File: src/core/sys/posix/signal.d [hunk 1]
```diff
@@ -544,6 +544,28 @@ version( CRuntime_Glibc )
         void function() sa_restorer;
     }
 }
+version( CRuntime_Musl )
```
Comment (rikkimax) [inline src/core/sys/posix/signal.d]: You missed an else.

Causing: ``src/core/sys/posix/signal.d(686): Error: static assert  "Architecture not supported."``.

Comment (rikkimax) [inline src/core/sys/posix/signal.d]: You missed an else.

Causing: ``src/core/sys/posix/signal.d(686): Error: static assert  "Architecture not supported."``.

**Comment by ZombineDev** [comment]
@yshui Thanks for looking into porting adding musl libc support to druntime. It looks like there are many things that need to be changed, which is why I highly recommend splitting your work into many commits. That would allow us merge changes that are ready much faster, so we can focus on the rest.

@wilzbach If I remember correctly, you had a tool to split work into multiple commits / pull-requests automatically?

**Comment by wilzbach** [comment]
This is really cool - any chance we can test this e.g. on CircleCi automatically? :)

> @wilzbach If I remember correctly, you had a tool to split work into multiple commits / pull-requests automatically?

Yes, it's a stupid bash script that will simply make a PR for every file.

https://github.com/wilzbach/splitup-prs

**Comment by ZombineDev** [comment]
@yshui I'm now reviewing your changes file by file and I'll get back to you with more comments after 1-2 hours.

> This is really cool - any chance we can test this e.g. on CircleCi automatically? :)

@wilzbach very good idea, but I don't think it's easily doable in the short term, since in order to to run the dmd/druntime/phobos tests on a musl libc system, we first need to compile them for which we need either cross-compilation (telling dmd that host != target, something which only LDC and GDC support well) or a good bootstrap process (but I'm not sure how usable are the dmd-cxx branches).

**Comment by wilzbach** [comment]
> good bootstrap process (but I'm not sure how usable are the dmd-cxx branches).

The latest C++ release can still compile master without any work. @ibuclaw's dmd-cxx branch contains a couple of backports, but IIRC druntime doesn't build with them yet.
Anyhow there's @CyberShadow's amazing [Digger](https://github.com/CyberShadow/Digger) which does an excellent job at bootstrapping.

**Comment by ibuclaw** [comment]
> @ibuclaw's dmd-cxx branch contains a couple of backports, but IIRC druntime doesn't build with them yet.

https://github.com/dlang/dmd/blob/dmd-cxx/travis.sh

I wouldn't have uncommented it if didn't work for me.

Someone should take ownership in looking at the glue/backend in that branch anyway. I have repeatedly said that only the frontend backed by gdc has been rigorously tested, which is backed by semaphore and buildbot ci which checks 6 distinct CPU targets.

File: src/core/sys/posix/sys/types.d [hunk 2]
```diff
@@ -604,6 +637,37 @@ version (CRuntime_Glibc)
 
     alias c_ulong pthread_t;
 }
+else version( CRuntime_Musl )
+{
+    union pthread_attr_t
```
Comment (ZombineDev) [inline src/core/sys/posix/sys/types.d]: Again, please separate the definitions per architecture. For reference:
http://git.musl-libc.org/cgit/musl/tree/arch/i386/bits/alltypes.h.in
http://git.musl-libc.org/cgit/musl/tree/arch/x86_64/bits/alltypes.h.in
http://git.musl-libc.org/cgit/musl/tree/arch/arm/bits/alltypes.h.in
http://git.musl-libc.org/cgit/musl/tree/arch/aarch64/bits/alltypes.h.in
http://git.musl-libc.org/cgit/musl/tree/arch/mipsn32/bits/alltypes.h.in
http://git.musl-libc.org/cgit/musl/tree/arch/mips64/bits/alltypes.h.in
...

**Comment by joakim-noah** [comment]
@yshui, nice work, I'll take a look.

@ZombineDev, never messed with Musl, can it be used on a regular glibc distro too?  If so, setting up CI should be easy.

@ibuclaw, I'd be happy to try testing the dmd-cxx branch, but I was under the impression you were going to update the frontend on that branch further, dlang/dmd#7227.

**Comment by ZombineDev** [comment]
> @ZombineDev, never messed with Musl, can it be used on a regular glibc distro too? If so, setting up CI should be easy.

@joakim-noah I wasn't sure too, but looking at https://github.com/emk/rust-musl-builder it seems it should be doable.

**Comment by wilzbach** [comment]
> @ZombineDev, never messed with Musl, can it be used on a regular glibc distro too? If so, setting up CI should be easy.

Yes, it actually looks pretty straight-forward. They even provide a gcc wrapper and just statically link it in:

https://wiki.musl-libc.org/getting-started.html#Using-the-musl-gcc-wrapper

(I haven't played with musl either)

File: src/core/time.d [hunk 3]
```diff
@@ -292,7 +292,7 @@ else version(Darwin) enum ClockType
     precise = 3,
     second = 6,
 }
-else version(linux) enum ClockType
+else version(CRuntime_Glibc) enum ClockType
```
Comment (joakim-noah) [inline src/core/time.d]: This is going to break this module for Bionic and the tests below for Musl.  I see no reason to change this, even if these constants aren't defined for you.

File: src/core/time.d [hunk 4]
```diff
@@ -355,9 +364,11 @@ version(Posix)
             case coarse: return CLOCK_MONOTONIC_COARSE;
             case normal: return CLOCK_MONOTONIC;
             case precise: return CLOCK_MONOTONIC;
-            case processCPUTime: return CLOCK_PROCESS_CPUTIME_ID;
             case raw: return CLOCK_MONOTONIC_RAW;
-            case threadCPUTime: return CLOCK_THREAD_CPUTIME_ID;
+            version(CRuntime_Glibc) {
+                case processCPUTime: return CLOCK_PROCESS_CPUTIME_ID;
+                case threadCPUTime: return CLOCK_THREAD_CPUTIME_ID;
+            }
```
Comment (joakim-noah) [inline src/core/time.d]: These won't be defined for Bionic anymore, and [they appear to be defined for Musl](http://git.musl-libc.org/cgit/musl/tree/include/time.h?h=v1.1.18#n87)?

File: src/core/stdc/stdio.d [hunk 5]
```diff
@@ -383,6 +409,16 @@ else version( CRuntime_Glibc )
...
+else version( CRuntime_Musl )
+{
+    struct fpos_t {
+        char[16] __opaque;
+        double __align;
+    }
+    struct _IO_FILE;
+    alias _IO_FILE _iobuf;
+    alias shared(_IO_FILE) FILE;
+}
```
[the remark this answers is not in the document]
Comment (ibuclaw) [reply src/core/stdc/stdio.d]: Strictly speaking we don't really need to define the structures of any `FILE` type.  But for completion sake, and to match other runtimes it probably should be added.

[the remark this answers is not in the document]
Comment (yshui) [reply src/core/stdc/stdio.d]: @ZombineDev _iobuf is defined because it's used in phobos.

**Comment by joakim-noah** [comment]
Do you plan on resurrecting the parts of this pull that haven't been submitted elsewhere? If so, maybe you can just reopen this pull with that remaining patch.

**Pull request closed without merging**